### PR TITLE
Added DeviceAuth and ExchangeAuth

### PR DESCRIPTION
**Pull request by ThisNils**
Added methods to login using ExchangeAuth and DeviceAuth. This bypasses the captcha invalid error. Data of both can be provided in an async function eg to get it from a database

### CODE REVIEW

File: example-test/index.js [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
 
 // Creation of the Client, autokill will kill the session when client is disposed. 
 // If you are not giving any tokens it will use the default ones that are needed.
 const client = new Client({
   email: '',
   password: '',
-  launcherToken: 'MzRhMDJjZjhmNDQxNGUyOWIxNTkyMTg3NmRhMzZmOWE6ZGFhZmJjY2M3Mzc3NDUwMzlkZmZlNTNkOTRmYzc2Y2Y=',
-  fortniteToken: 'ZWM2ODRiOGM2ODdmNDc5ZmFkZWEzY2IyYWQ4M2Y1YzY6ZTFmMzFjMjExZjI4NDEzMTg2MjYyZDM3YTEzZmM4NGQ=',
-  autokill: true,
+  deviceAuth: () => JSON.parse(fs.readFileSync('./deviceauth.json', 'utf-8')),
```
Comment (iXyles) [inline example-test/index.js]: Should be an async function...

File: example-test/index.js [hunk 2]
```diff
@@ -18,6 +18,7 @@ const communicator = new Communicator(client);
 (async () => {
   // Perform the login process of the "client"
 
+  client.authenticator.on('device_auth_created', d => fs.writeFileSync('./deviceauth.json', JSON.stringify(d)));
```
Comment (iXyles) [inline example-test/index.js]: Same goes here for async

File: src/Client/Authenticator.js [hunk 3]
```diff
@@ -18,8 +18,57 @@ module.exports = class Authenticator extends EventEmitter {
...
    */
   async login() {
-    const token = await this.getOAuthToken();
-    if (token.error) return token;
+    if(this.client.createNewDeviceAuth) {
+      const deviceauth = await this.createDeviceAuth();
+      if(!deviceauth.success) return deviceauth.response;
+    }
+
+    let token;
```
Comment (iXyles) [inline src/Client/Authenticator.js]: So why is this on a unique line? token is still only set once.

File: src/Client/Authenticator.js [hunk 4]
```diff
@@ -56,6 +105,47 @@ module.exports = class Authenticator extends EventEmitter {
...
+      includePerms: true,
+      token_type: 'eg1',
+    };
+
+    const res = await this.client.requester.sendPost(false, Endpoints.OAUTH_TOKEN, `basic ${this.client.iosToken}`, exchangeData, {'Content-Type': 'application/x-www-form-urlencoded'}, true);
+    if (res.error) return {success: false, response: res.error};
+
+    const deviceAuthDetails = await this.client.requester.sendPost(false, `${Endpoints.DEVICE_AUTH}/${res.account_id}/deviceAuth`, `bearer ${res.access_token}`);
+    if(deviceAuthDetails.error) return {success: false, response: deviceAuthDetails.error};
+    console.log(deviceAuthDetails);
```
Comment (iXyles) [inline src/Client/Authenticator.js]: logger should not be in the code

File: src/Client/Authenticator.js [hunk 5]
```diff
@@ -56,6 +105,47 @@ module.exports = class Authenticator extends EventEmitter {
     return { success: true }; // successful login
   }
 
+  async createDeviceAuth() {
```
Comment (iXyles) [inline src/Client/Authenticator.js:88]: Missing comments

File: src/Client/Authenticator.js [hunk 6]
```diff
@@ -56,6 +105,47 @@ module.exports = class Authenticator extends EventEmitter {
...
+    this.client.deviceAuth = {
+      deviceId: deviceAuthDetails.deviceId,
+      accountId: deviceAuthDetails.accountId,
+      secret: deviceAuthDetails.secret
+    }
+    this.emit('device_auth_created', this.client.deviceAuth);
+    return {success: true, response: this.client.deviceAuth};
+  }
+
+  async deleteAllDeviceAuths(dontDeleteIds = []) {
```
Comment (iXyles) [inline src/Client/Authenticator.js]: Missing comments

File: src/Client/Authenticator.js [hunk 7]
```diff
@@ -18,8 +18,57 @@ module.exports = class Authenticator extends EventEmitter {
    * @return {object} JSON Object of login result
    */
   async login() {
```
Comment (iXyles) [inline src/Client/Authenticator.js]: Comment needs to be re-phrased.

File: src/Client/Authenticator.js [hunk 8]
```diff
@@ -243,7 +335,7 @@ module.exports = class Authenticator extends EventEmitter {
       `${Endpoints.ENTITLEMENTS}/${login.account_id}/entitlements?start=0&count=5000`,
       `bearer ${login.access_token}`,
     );
-
+    console.log(this.entitlements)
```
Comment (iXyles) [inline src/Client/Authenticator.js]: logger should not be in code.

**Review by iXyles — CHANGES_REQUESTED** [review]
Except for comments, the code need to follow the rules set by .eslintrc.js with eslint.

File: example-test/index.js (hunk 1 above)
Comment (iXyles) [reply example-test/index.js]: Also, why remove the example code that was there previously?

File: src/Client/Authenticator.js [hunk 9]
```diff
@@ -18,8 +18,57 @@ module.exports = class Authenticator extends EventEmitter {
...
+      this.setAuthData(token);
+      if(this.client.deleteOtherDeviceAuths) {
+        const deleteOthers = await this.deleteAllDeviceAuths([deviceAuthDetails.deviceId]);
+        if(deleteOthers.error) return deleteOthers;
+      }
+      // Setup kill hook for the session token - to prevent login issues on restarts
+      if (this.client.autoKillSession && !this.killHook) {
+        exitHook(async (callback) => {
+          // eslint-disable-next-line no-console
+          await console.info(await this.killCurrentSession());
```
Comment (ghost) [inline src/Client/Authenticator.js]: The first await has no effect and can be removed.

**Comment by iXyles** [comment]
Based a new branch of commit '162b136' and it is now merge into the master. Closing this PR since it is already in the master.

**Pull request closed without merging**